nntrainer/layers/pooling2d_layer.cpp

```diff
@@ -337,17 +359,47 @@ Tensor Pooling2DLayer::pooling2d(Tensor &in, bool training, Tensor &output) {
     };
     break;
   }
-  case PoolingType::average:
-  case PoolingType::global_max:
+  case PoolingType::global_max: {
+    break;
+  }
   case PoolingType::global_average:
+  case PoolingType::average: {
+    pool_fn = [&, this](const float *in_data, int start_h, int start_w) {
+      int end_h = start_h + patch_height;
+      int end_w = start_w + patch_width;
+      float total = 0.0f;
+      int cnt = 0;
+
+      for (int h = start_h; h < end_h; ++h) {
+        if (h < 0 || in_height <= h) {
+          continue;
+        }
+        for (int w = start_w; w < end_w; ++w) {
+          if (w < 0 || in_width <= w) {
+            continue;
+          }
+          float val = in_data[h * in_width + w];
+          cnt++;
+          total += val;
+        }
+      }
+
+      if (training) {
+        max_idx.push_back(cnt);
```

Comment:
max_idx being pushed in average pooling?

Reply:
looks like max_idx stores the effective average counter. Can you please explain what is effective average counter?

Reply:
this is to follow the semantics with tensorflow, I found that, if padding is part of the patch, it shouldn't be part of average calculation.

eg) pooling of below
x x x
x 3 3
x 3 3 
= 12 / 4 = 3

This counter has to be remembered or recalculated when backwarding as well.

Reply:
Sounds reasonable. Please add some explanations in the comments too then.

Reply:
Done within new commit